projects/graph/graph.py

```diff
@@ -1,55 +1,160 @@
 """
 Simple graph implementation
 """
+
 from util import Stack, Queue  # These may come in handy
 
 class Graph:
     """Represent a graph as a dictionary of vertices mapping labels to edges."""
     def __init__(self):
+
         self.vertices = {}
+        
     def add_vertex(self, vertex):
         """
         Add a vertex to the graph.
-        """
-        pass  # TODO
+        """  
+
+        # TODO
+        self.vertices[vertex] = set()
+
     def add_edge(self, v1, v2):
         """
         Add a directed edge to the graph.
         """
-        pass  # TODO
+
+        # TODO
+        if v1 in self.vertices and v2 in self.vertices:
+            self.vertices[v1].add(v2)
+        else:
+            raise IndexError("That vertex does not exist!")
+
     def bft(self, starting_vertex):
         """
         Print each vertex in breadth-first order
         beginning from starting_vertex.
         """
-        pass  # TODO
+
+        # TODO
+        # Create an empty set to store visited nodes
+        visited = set()
+        # Create an empty Queue and enqueue the starting vertex
+        q = Queue()
+        q.enqueue(starting_vertex)
+        # While the queue is not empty...
+        while q.size() > 0:
+            # Dequeue the first vertex
+            v = q.dequeue()
+            # If that vertex has not been visited...
+            if v not in visited:
+                # Mark it as visited
+                visited.add(v)
+                print(v)
+                # Then add all of its neighbors to the back of the queue
+                for neighbor in self.vertices[v]:
+                    q.enqueue(neighbor)
+
     def dft(self, starting_vertex):
         """
         Print each vertex in depth-first order
         beginning from starting_vertex.
         """
-        pass  # TODO
+
+        # TODO
+        # Create an empty set to store visited nodes
+        visited = set()
+        # Create an empty Stack and push the starting vertex
+        s = Stack()
+        s.push(starting_vertex)
+        # While the stack is not empty...
+        while s.size() > 0:
+            # Pop the first vertex
+            v = s.pop()
+            # If that vertex has not been visited...
+            if v not in visited:
+                # Mark it as visited
+                visited.add(v)
+                print(v)
+                # Then add all of its neighbors to the top of the stack
+                for neighbor in self.vertices[v]:
+                    s.push(neighbor)
+
     def dft_recursive(self, starting_vertex):
         """
         Print each vertex in depth-first order
         beginning from starting_vertex.
         This should be done using recursion.
         """
```

Comment:
Good job on this so far. I don't know if you still need to get this recursive part done but if not here is a hint - 

This recursive function is very similar to how you made the first dft. You will create a set of your visited nodes, make a for loop and check to see if the node you're looking at is not in the set, and if it is not in the set then you call the recursive function again.